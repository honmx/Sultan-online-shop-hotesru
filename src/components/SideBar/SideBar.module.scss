@import "../../scss/variables.scss";

.sidebar {}

.sidebar>*:not(:last-child) {
  margin-bottom: 20px;
}

.titleContainer {
  display: flex;
  align-items: center;
}

.title {
  font-size: 20px;
  margin-right: 10px;
}

.selector {
  margin: 20px 0;

  &:not(:last-child) {
    &:not(:first-child) {
      position: relative;
      padding-bottom: 20px;

      &::after {
        content: "";
        position: absolute;
        bottom: 0;
        left: 0;
        width: 200px;
        border: 1px dotted rgba($color: $primaryGrey, $alpha: 0.2);
      }
    }
  }
}

.buttonsContainer {
  display: flex;
}

.show {
  margin-right: 10px;
}

.filterbar {
  flex-direction: column;
  align-items: start;
}

.filterbar>*:not(:last-child) {
  margin-bottom: 5px;
}

@media screen and (max-width: 1023px) {
  .title {
    font-size: 14px;
  }
}