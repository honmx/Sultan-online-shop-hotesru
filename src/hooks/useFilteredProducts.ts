import { useMemo } from "react";
import { IProduct } from "../types/IProducts";
import { sortOptionsType } from "../types/SortOptionsType";
import { sortBy } from "../helpers/sortBy";

interface IDependecies {
  sort: keyof sortOptionsType;
  selectedFilters: string[];
  priceRange: { min: number; max: number };
  selectedProducers: string[];
  selectedBrands: string[];
}

export const useFilteredProducts = (products: IProduct[], { sort, selectedFilters, priceRange, selectedProducers, selectedBrands }: IDependecies) => {
  const filteredProducts: IProduct[] = useMemo(() => {
    if (!products) return [];

    return sortBy(sort, products)
      .filter(product => {
        return selectedFilters.length === 0
          ? true
          : product.type.subtypes?.some(elem => selectedFilters.includes(elem));
      })
      .filter(product => {
        return product.price.value >= priceRange.min && product.price.value <= priceRange.max;
      })
      .filter(product => {
        return selectedProducers.length === 0
          ? true
          : selectedProducers.includes(product.producer);
      })
      .filter(product => {
        return selectedBrands.length === 0
          ? true
          : selectedBrands.includes(product.brand);
      });
  
    }, [products, sort, selectedFilters, selectedProducers, selectedBrands, priceRange.min, priceRange.max]);
    
  return [filteredProducts];
}