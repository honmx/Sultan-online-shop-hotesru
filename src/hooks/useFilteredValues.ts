import { useEffect, useState } from "react";

interface IDependencies {
  inputValue: string;
}

export const useFilteredValues = (values: string[], { inputValue }: IDependencies) => {
  const [setOfFilteredValues, setSetofFilteredValues] = useState(Array.from(new Set(values)).sort((a, b) => a.localeCompare(b)))
  const [filtered, setFiltered] = useState<string[]>(setOfFilteredValues);

  useEffect(() => {
    setFiltered(setOfFilteredValues.filter(value => value.toLowerCase().startsWith(inputValue.toLowerCase())));
  }, [inputValue]);

  useEffect(() => {
    setFiltered(Array.from(new Set(values)).sort((a, b) => a.localeCompare(b)));
  }, [values]);

  return [filtered];
}