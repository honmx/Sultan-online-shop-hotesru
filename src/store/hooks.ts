import { useDispatch, useSelector } from "react-redux";
import { TypedUseSelectorHook } from "react-redux/es/types";
import { AppDispatch, RootState } from "./store";

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
