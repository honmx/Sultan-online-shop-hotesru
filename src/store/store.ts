import { configureStore, getDefaultMiddleware } from "@reduxjs/toolkit";
import { apiSlice } from "./slices/apiSlice";
import cartSlice from "./slices/cartSlice";
import categorySlice from "./slices/categorySlice";
import productsSlice from "./slices/productsSlice";

const store = configureStore({
  reducer: {
    products: productsSlice,
    category: categorySlice,
    cart: cartSlice,
    // [apiSlice.reducerPath]: apiSlice.reducer,
  },
  // middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiSlice.middleware)
});

export { store };
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;