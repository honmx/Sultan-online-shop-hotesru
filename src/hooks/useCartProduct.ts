import { useAppSelector } from "../store/hooks"
import { IProduct } from "../types/IProducts";

export const useCartProduct = (product: IProduct | undefined) => {
  const cartProduct = useAppSelector(state => state.cart.cartItems).find(item => item.product.barcode === product?.barcode);

  return { inCart: Boolean(cartProduct) }
}