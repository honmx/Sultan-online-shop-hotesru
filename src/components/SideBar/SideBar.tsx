import React, { FC, useState } from "react";
import { IProduct } from "../../types/IProducts";
import Button from "../UI/Button/Button";
import RangeSelector from "../UI/RangeSelector/RangeSelector";
import SelectSelector from "../UI/SelectSelector/SelectSelector";
import Title from "../UI/Title/Title";
import trash from "../../assets/trash.svg";
import FilterBar from "../FilterBar/FilterBar";
import { useAppDispatch, useAppSelector } from "../../store/hooks";
import { filters } from "../../helpers/data/filters";
import { FilterType } from "../../types/FilterType";
import { useSmallerDevice } from "../../hooks/useSmallerDevice";
import arrowDown from "../../assets/arrow-down.svg";
import arrowLeft from "../../assets/arrow-left.svg";
import s from "./SideBar.module.scss";
import { clearAllFilters, setBrands, setPrice, setProducers } from "../../store/slices/categorySlice";

interface Props {
  className?: string;
  products: IProduct[];
  categoryType: FilterType;
}

const SideBar: FC<Props> = ({ className, products, categoryType }) => {

  const dispatch = useAppDispatch();
  
  const selectedFilters = useAppSelector(state => state.category.filters);

  const [min, setMin] = useState<number>(0);
  const [max, setMax] = useState<number>(10000);

  const [selectedProducers, setSelectedProducers] = useState<string[]>([]);
  const [selectedBrands, setSelectedBrands] = useState<string[]>([]);

  const [active, setActive] = useState<boolean>(false);

  const [smallerDevice] = useSmallerDevice(1023);

  const applyFilters = () => {
    dispatch(setPrice({min, max}));
    dispatch(setProducers(selectedProducers));
    dispatch(setBrands(selectedBrands));
  }

  const clearFilters = () => {
    setMin(0);
    setMax(10000);
    setSelectedProducers([]);
    setSelectedBrands([]);
  }

  return (
    <div className={`${s.sidebar} ${className}`}>
      <div className={s.titleContainer}>
        <Title usualText="Подбор по параметрам" className={s.title} />
        {
          smallerDevice &&
          <Button
            img={active ? arrowLeft : arrowDown}
            w={smallerDevice ? "30px" : "40px"}
            p={"0px"}
            ar={1}
            onClick={() => setActive(prev => !prev)}
          />
        }
      </div>
      {
        (!smallerDevice || active) &&
        <div className={s.selectors}>
          <RangeSelector className={s.selector} name="Цена ₸" min={min} max={max} setMin={setMin} setMax={setMax} />
          <SelectSelector
            className={s.selector}
            name="Производитель"
            values={products.map(product => product.producer)}
            selected={selectedProducers}
            setSelected={setSelectedProducers}
          />
          <SelectSelector
            className={s.selector}
            name="Бренд"
            values={products.map(product => product.brand)}
            selected={selectedBrands}
            setSelected={setSelectedBrands}
          />
          <div className={s.buttonsContainer}>
            <Button className={s.show} onClick={applyFilters}>Показать</Button>
            <Button img={trash} p={"18.5px"} onClick={clearFilters}></Button>
          </div>
        </div>
      }
      {
        !smallerDevice &&
        <FilterBar className={s.filterbar} filters={filters[categoryType]} selectedFilters={selectedFilters} />
      }
    </div>
  )
};

export default SideBar;
