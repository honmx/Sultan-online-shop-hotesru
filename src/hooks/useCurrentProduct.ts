import { useAppSelector } from "../store/hooks";

export const useCurrentProduct = (id: number) => {
  const products = useAppSelector(state => state.products.products);
  const product = products.find(item => item.id === id);

  return product;
} 