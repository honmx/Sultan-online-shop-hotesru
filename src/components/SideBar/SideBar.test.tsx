import { fireEvent, getByTestId, render, screen } from "@testing-library/react";
import SideBar from "./SideBar";
import * as typedReduxHooks from "../../store/hooks";

jest.mock("react-redux");
const mockedUseSelector = jest.spyOn(typedReduxHooks, "useAppSelector");
const mockedUseDispatch = jest.spyOn(typedReduxHooks, "useAppDispatch");

describe("sidebar tests", () => {
  test("sidebar renders", () => {
    mockedUseDispatch.mockReturnValue(jest.fn());
    mockedUseSelector.mockReturnValue([]);
    
    const component = render(<SideBar products={[]} categoryType={"cosmetics-and-hygiene"} />)
    
    expect(component).toMatchSnapshot();
  });
  
  test("applies all filters", () => {
    const dispatch = jest.fn();
    mockedUseDispatch.mockReturnValue(dispatch);
    mockedUseSelector.mockReturnValue([]);

    const { getByTestId } = render(<SideBar products={[]} categoryType={"cosmetics-and-hygiene"} />);
  
    const applyBtn = getByTestId("applybtn");

    fireEvent.click(applyBtn);

    expect(dispatch).toHaveBeenCalled();
  });
  
  test("clears all filters", () => {
    mockedUseDispatch.mockReturnValue(jest.fn());
    mockedUseSelector.mockReturnValue([]);

    const { getByTestId } = render(<SideBar products={[]} categoryType={"cosmetics-and-hygiene"} />);
  
    const OnClickFn = jest.fn();

    const applyBtn = getByTestId("clearbtn");
    applyBtn.onclick = OnClickFn; 

    fireEvent.click(applyBtn);

    expect(OnClickFn).toHaveBeenCalled();
  });
});