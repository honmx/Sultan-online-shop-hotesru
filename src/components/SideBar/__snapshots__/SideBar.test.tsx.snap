// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar tests sidebar renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sidebar undefined"
      >
        <div
          class="titleContainer"
        >
          <div
            class="title title"
          >
            <span
              class="accent"
            >
               
            </span>
            <span
              class="usual"
            >
              Подбор по параметрам
            </span>
          </div>
        </div>
        <div
          class="selectors"
        >
          <div
            class="rangeContainer selector"
          >
            <p
              class="name"
            >
              Цена ₸
            </p>
            <div
              class="inputContainer"
            >
              <input
                type="text"
                value="0"
              />
              <span>
                -
              </span>
              <input
                type="text"
                value="10000"
              />
            </div>
          </div>
          <div
            class="selectContainer selector"
          >
            <p
              class="title"
            >
              Производитель
            </p>
            <input
              class="undefined input"
              placeholder="Поиск..."
              type="text"
              value=""
            />
            <form
              action=""
              class="form"
            />
          </div>
          <div
            class="selectContainer selector"
          >
            <p
              class="title"
            >
              Бренд
            </p>
            <input
              class="undefined input"
              placeholder="Поиск..."
              type="text"
              value=""
            />
            <form
              action=""
              class="form"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="show button"
              data-testid="applybtn"
            >
              Показать
            </button>
            <button
              class="undefined button"
              data-testid="clearbtn"
              style="padding: 18.5px;"
            >
              <img
                src="trash.svg"
                style="margin-left: 0px;"
              />
            </button>
          </div>
        </div>
        <div
          class="filters filterbar"
        >
          <button
            class="button"
          >
            Уход за телом
          </button>
          <button
            class="button"
          >
            Уход за руками
          </button>
          <button
            class="button"
          >
            Уход за ногами
          </button>
          <button
            class="button"
          >
            Уход за лицом
          </button>
          <button
            class="button"
          >
            Уход за волосами
          </button>
          <button
            class="button"
          >
            Средства для загара
          </button>
          <button
            class="button"
          >
            Средства для бритья
          </button>
          <button
            class="button"
          >
            Подарочные наборы
          </button>
          <button
            class="button"
          >
            Гигиеническая продукция
          </button>
          <button
            class="button"
          >
            Гигиена полости рта
          </button>
          <button
            class="button"
          >
            Бумажная продукция
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sidebar undefined"
    >
      <div
        class="titleContainer"
      >
        <div
          class="title title"
        >
          <span
            class="accent"
          >
             
          </span>
          <span
            class="usual"
          >
            Подбор по параметрам
          </span>
        </div>
      </div>
      <div
        class="selectors"
      >
        <div
          class="rangeContainer selector"
        >
          <p
            class="name"
          >
            Цена ₸
          </p>
          <div
            class="inputContainer"
          >
            <input
              type="text"
              value="0"
            />
            <span>
              -
            </span>
            <input
              type="text"
              value="10000"
            />
          </div>
        </div>
        <div
          class="selectContainer selector"
        >
          <p
            class="title"
          >
            Производитель
          </p>
          <input
            class="undefined input"
            placeholder="Поиск..."
            type="text"
            value=""
          />
          <form
            action=""
            class="form"
          />
        </div>
        <div
          class="selectContainer selector"
        >
          <p
            class="title"
          >
            Бренд
          </p>
          <input
            class="undefined input"
            placeholder="Поиск..."
            type="text"
            value=""
          />
          <form
            action=""
            class="form"
          />
        </div>
        <div
          class="buttonsContainer"
        >
          <button
            class="show button"
            data-testid="applybtn"
          >
            Показать
          </button>
          <button
            class="undefined button"
            data-testid="clearbtn"
            style="padding: 18.5px;"
          >
            <img
              src="trash.svg"
              style="margin-left: 0px;"
            />
          </button>
        </div>
      </div>
      <div
        class="filters filterbar"
      >
        <button
          class="button"
        >
          Уход за телом
        </button>
        <button
          class="button"
        >
          Уход за руками
        </button>
        <button
          class="button"
        >
          Уход за ногами
        </button>
        <button
          class="button"
        >
          Уход за лицом
        </button>
        <button
          class="button"
        >
          Уход за волосами
        </button>
        <button
          class="button"
        >
          Средства для загара
        </button>
        <button
          class="button"
        >
          Средства для бритья
        </button>
        <button
          class="button"
        >
          Подарочные наборы
        </button>
        <button
          class="button"
        >
          Гигиеническая продукция
        </button>
        <button
          class="button"
        >
          Гигиена полости рта
        </button>
        <button
          class="button"
        >
          Бумажная продукция
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
